File: C++풀이/문자열압축.cpp
#include <string>
#include <vector>
using namespace std;

int solution(string s) {
    int sz = s.length();
    int answer = sz;
    
    // 1 ~ n/2 개수만큼 나눌 수 있다.
    for(int i = 1; i < (sz/2)+1; i++){
        string prev, l = "";
        int cnt = 1;
        
        prev = s.substr(0,i);
        for(int j = i; j < sz; j += i){
            string next = s.substr(j,i);
            if(prev == next) { cnt++;}
            else{
                if((i == 1 && cnt == 1) || cnt == 1){
                    l.append(prev);
                }
                else{
                    l.append(to_string(cnt)+prev);
                    
                }
                prev = next;
                cnt = 1;
            }
        }
        if(cnt > 1) l.append(to_string(cnt)+prev);
        else l.append(prev);
        answer = answer > l.length()? l.length() : answer;
    }
    
    return answer;
}

File: JAVA풀이/문자열압축.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class 문자열압축{

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String s = br.readLine();

        int answer = -1;
        for(int i = 1; i <= s.length()/2; i++){
            int len = 0;

            // init
            int j = i, cnt = 1;
            String prev = s.substring(0,j), cur = "";

            while(j < s.length()){
                int end = (j+i) >= s.length() ? s.length() : (j+i);
                cur = s.substring(j,end);
                j = end;

                if(cur.equals(prev)){ // 같으면 카운터만 올리기
                    cnt++;
                } else{ // 같지 않으면 추가
                    len += findCount(cnt, i);
                    if(answer != -1 && len > answer){ break; }
                    cnt = 1;
                }
                prev = cur;
            }
            // 마지막 부분 저장
            len += findCount(cnt, cur.length());

            // answer
            if(answer == -1 || answer > len){
                answer = len;
            }
        }

        System.out.println(answer);
    }
    static int findCount(int cnt, int len){
        if(cnt == 1) return len;
        else{
            String c = Integer.toString(cnt);
            return c.length() + len;
        }
    }
}